File: e2e_tests/python/server_clients/stdio_server.py
import logging
import os
import shutil
from typing import Any

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client

from server_clients.server import Server


class StdioServer(Server):
    """Manages MCP server connections and tool execution for servers running locally over stdio."""

    def __init__(self, name: str, config: dict[str, Any]) -> None:
        super().__init__(name, config)

    async def initialize(self) -> None:
        """Initialize the server connection."""
        command = (
            shutil.which("npx")
            if self.config["command"] == "npx"
            else self.config["command"]
        )
        if command is None:
            raise ValueError("The command must be a valid string and cannot be None.")

        server_params = StdioServerParameters(
            command=command,
            args=self.config["args"],
            env=(
                {**os.environ, **self.config["env"]} if self.config.get("env") else None
            ),
        )
        try:
            self._client = stdio_client(server_params)
            self.read, self.write = await self._client.__aenter__()
            session = ClientSession(self.read, self.write)
            self.session = await session.__aenter__()
            await self.session.initialize()
        except Exception as e:
            logging.exception(f"Error initializing server {self.name}: {e}")
            raise

File: e2e_tests/typescript/src/server_clients/stdio_server.ts
import {
  StdioClientTransport,
  StdioServerParameters,
} from "@modelcontextprotocol/sdk/client/stdio.js";
import { Server } from "./server.js";
import logger from "../logger.js";

/**
 * Manages MCP server connections and tool execution for servers running locally over stdio.
 */
export class StdioServer extends Server {
  /**
   * Initialize the server connection.
   * @throws ValueError if initialization parameters are invalid
   * @throws RuntimeError if server fails to initialize
   */
  async initialize(): Promise<void> {
    if (!this.config.command) {
      throw new Error("The command must be a valid string and cannot be None.");
    }

    const serverParams: StdioServerParameters = {
      command: this.config.command,
      args: this.config.args,
      env: this.config.env ? { ...process.env, ...this.config.env } : undefined,
    };

    const transport = new StdioClientTransport(serverParams);

    try {
      await this.client.connect(transport);
    } catch (e) {
      logger.error(`Error initializing server ${this.name}: ${e}`);
      throw e;
    }
  }
}
